File: bench/echo.ml
open Riot

module Echo = struct
  module Server = struct
    open Atacama.Handler
    include Atacama.Handler.Default

    type state = int
    type error = [ `noop ]

    let pp_err fmt `noop = Format.fprintf fmt "Noop"

    let handle_data data socket state =
      match Atacama.Connection.send socket data with
      | Ok _bytes -> Continue (state + 1)
      | Error _ -> Close state
  end

  let start () =
    Logger.set_log_level (Some Info);
    Atacama.start_link ~port:2113 (module Server) 0
end

let () = Riot.start ~apps:[ (module Logger); (module Echo) ] ()

File: bench/echo.exs
Mix.install([:thousand_island])

require Logger

defmodule Echo do
  use ThousandIsland.Handler

  @impl ThousandIsland.Handler
  def handle_data(data, socket, state) do
    ThousandIsland.Socket.send(socket, data)
    {:continue, state}
  end
end

{:ok, pid} = ThousandIsland.start_link(port: 2112, handler_module: Echo)

Process.sleep(:infinity)
